lib/rules/curly.js

```diff
@@ -114,36 +114,96 @@ module.exports = function(context) {
+     *   `null` if it doesn't matter, depending on the rule options. It can be modified to change
+     *   the final behavior of "check".
+     *   "check" will be a function reporting appropriate problems depending on the other
+     *   properties.
      */
-    function checkBody(node, body, name, suffix) {
+    function prepareCheck(node, body, name, suffix) {
         var hasBlock = (body.type === "BlockStatement");
+        var expected = null;
 
-        if (multiOnly) {
+        if (node.type === "IfStatement" && node.consequent === body && requiresBraceOfConsequent(node)) {
+            expected = true;
+        } else if (multiOnly) {
             if (hasBlock && body.body.length === 1) {
-                reportUnnecessaryBraceError(node, name, suffix);
+                expected = false;
             }
         } else if (multiLine) {
-            if (!hasBlock && !isCollapsedOneLiner(body)) {
-                reportExpectedBraceError(node, name, suffix);
+            if (!isCollapsedOneLiner(body)) {
+                expected = true;
             }
         } else if (multiOrNest) {
             if (hasBlock && body.body.length === 1 && isOneLiner(body.body[0])) {
-                reportUnnecessaryBraceError(node, name, suffix);
-            } else if (!hasBlock && !isOneLiner(body)) {
-                reportExpectedBraceError(node, name, suffix);
+                expected = false;
+            } else if (!isOneLiner(body)) {
+                expected = true;
             }
         } else {
-            if (!hasBlock) {
-                reportExpectedBraceError(node, name, suffix);
+            expected = true;
+        }
+
+        return {
+            actual: hasBlock,
+            expected: expected,
+            check: function() {
+                if (this.expected !== null && this.expected !== this.actual) {
+                    if (this.expected) {
+                        reportExpectedBraceError(node, name, suffix);
+                    } else {
+                        reportUnnecessaryBraceError(node, name, suffix);
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * Prepares to check the bodies of a "if", "else if" and "else" chain.
+     * @param {ASTNode} node The first IfStatement node of the chain.
+     * @returns {object[]} prepared checks for each body of the chain. See `prepareCheck` for more
+     *   information.
+     */
+    function prepareIfChecks(node) {
+        var preparedChecks = [];
+        do {
+            preparedChecks.push(prepareCheck(node, node.consequent, "if", "condition"));
+            if (node.alternate && node.alternate.type !== "IfStatement") {
+                preparedChecks.push(prepareCheck(node, node.alternate, "else"));
+                break;
             }
+            node = node.alternate;
+        } while (node);
+
+        if (consistent) {
+            var expected = preparedChecks.reduce(function(result, preparedCheck) {
```

Comment:
I'm not sure I understand here. The `reduce` will run for every if statement starting an if/elseif/else chain. Each chain should be independent, like the `quote-props` rule, so we can't use a global variable here. I will replace this by a `some` though, as it is simpler and more efficient.